rpc-broker/src/dbus.c

```diff
@@ -0,0 +1,423 @@
+*/
+
+#include "rpc-broker.h"
+
+
+DBusConnection *create_dbus_connection(void)
+{
+    DBusError error;
+    dbus_error_init(&error);
+    DBusConnection *conn = dbus_bus_get(DBUS_BUS_SYSTEM, &error);
+
+    if (dbus_error_is_set(&error))
+        DBUS_BROKER_WARNING("<DBus Connection Error> [%s]", error.message);
+
+    return conn;
+}
+
+struct dbus_broker_server *start_server(int port)
+{
+    struct dbus_broker_server *server = malloc(sizeof *server);
+    server->dbus_socket = socket(AF_INET, SOCK_STREAM, 0);
+    if (server->dbus_socket < 0)
+        DBUS_BROKER_ERROR("socket");
+
+    server->addr.sin_family = AF_INET;
+    server->addr.sin_addr.s_addr = INADDR_ANY;
+    server->addr.sin_port = htons(port);
+    server->peer.sin_family = AF_INET;
+    server->peer.sin_addr.s_addr = INADDR_ANY;
+    server->peer.sin_port = 0;
+
+    if (bind(server->dbus_socket, &server->addr, sizeof(server->addr)) < 0)
+        DBUS_BROKER_ERROR("bind");
+
+    if (listen(server->dbus_socket, 1) < 0)
+        DBUS_BROKER_ERROR("listen");
+
+    return server;
+}
+
+void dbus_default(struct dbus_message *dmsg)
+{
+    dmsg->destination = DBUS_DB_DEST;
+    dmsg->interface = DBUS_DB_IFACE;
+    dmsg->path = DBUS_BASE_PATH;
+    dmsg->arg_number = 1;
+    snprintf(dmsg->arg_sig, DBUS_MAX_ARG_LEN - 1, "s");
+}
+
+int connect_to_system_bus(void)
+{
+    int srv = socket(AF_UNIX, SOCK_STREAM, 0);
+    if (srv < 0)
+        DBUS_BROKER_ERROR("socket");
+
+    struct sockaddr_un addr;
+    addr.sun_family = AF_UNIX;
+    // could have bus-path override (LEN would need checking)
+    snprintf(addr.sun_path, DBUS_BUS_ADDR_LEN, DBUS_BUS_ADDR);
+
+    if (connect(srv, (struct sockaddr *) &addr, sizeof(addr)) < 0)
+        DBUS_BROKER_ERROR("connect");
+
+    return srv;
+}
+
+/*
+ * Used by client communications over port-5555 where, the raw-bytes are being
+ * read directly from the client file-descriptor.  The `char` buffer is
+ * de-marshalled into a dbus-protocol object `DBusMessage`
+ */
+signed int convert_raw_dbus(struct dbus_message *dmsg,
+                            const char *msg, size_t len)
+{
+    DBusError error;
+    dbus_error_init(&error);
+
+    DBusMessage *dbus_msg = NULL;
+    dbus_msg = dbus_message_demarshal(msg, len, &error);
+
+    if (dbus_error_is_set(&error)) {
+        DBUS_BROKER_WARNING("<De-Marshal failed> [Length: %zu] error: %s",
+                              len, error.message);
+        return -1;
+    }
+
+    const char *destination = dbus_message_get_destination(dbus_msg);
+    dmsg->destination = destination ? destination : "NULL";
+
+    const char *path = dbus_message_get_path(dbus_msg);
+    dmsg->path = path ? path : "/";
+
+    const char *interface = dbus_message_get_interface(dbus_msg);
+    dmsg->interface = interface ? interface : "NULL";
+
+    const char *member = dbus_message_get_member(dbus_msg);
+    dmsg->member = member ? member : "NULL";
+
+    int ret = dbus_message_get_type(dbus_msg);
+/*
+    // XXX
```

Comment:
Maybe instead of commenting out, this debug output should be in its own function and conditionally called if (verbose_logging)